C error: When using match in a field Option

### Describe the bug

When a field of a struct is OptionType(?int64) when trying to use the field in a match, an error occurs.



### Expected Behavior

The following text should be printed: "text: num: none"

### Current Behavior

```
C:/Users/AndreLuiz/AppData/Local/Temp/v_0/test.6182178452072019841.tmp.c:497: warning: WINVER redefined
C:/Users/AndreLuiz/AppData/Local/Temp/v_0/../../../../../..C:\v\vlib\builtin\builtin.c.v:68: warning: implicit declaration of function 'tcc_backtrace'
C:/Users/AndreLuiz/AppData/Local/Temp/v_0/../../../../../..C:\v\vlib\builtin\builtin_windows.c.v:288: warning: assignment from incompatible pointer type
C:/Users/AndreLuiz/AppData/Local/Temp/v_0/../../../../../..C:\v\vlib\builtin\builtin_windows.c.v:288: warning: assignment makes pointer from integer without a cast
C:/Users/AndreLuiz/AppData/Local/Temp/v_0/../../../../../..C:\v\vlib\builtin\builtin_windows.c.v:288: warning: cast between pointer and integer of different size
C:/Users/AndreLuiz/AppData/Local/Temp/v_0/../../../../../..C:\Users\AndreLuiz\Desktop\test.v:12: error: ')' expected (got "_t2")
builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```



### Reproduction Steps

Code:
https://play.vlang.io/p/d83b8ae89e
https://devbits.app/p/wtRy53A1m8jx


```v
pub struct ParamPrecos {
pub:
	codigo string [required]
	tabela_referencia ?i64
}

fn main() {
	
	pp := ParamPrecos{codigo: "opa"}
		
	tx_ref := match pp.tabela_referencia {
		none { "num: none" }
		else { "num: ${pp.tabela_referencia?}" }
	}
	
	println("text: ${tx_ref}")
	
}
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.3 6ff1c0a

### Environment details (OS name and version, etc.)

V full version: V 0.3.3 d7a418f.6ff1c0a
OS: windows, Microsoft Windows 11 Pro Insider Preview v25324 64-bit
Processor: 8 cpus, 64bit, little endian,

getwd: C:\Users\AndreLuiz
vexe: C:\v\v.exe
vexe mtime: 2023-03-31 13:39:59

vroot: OK, value: C:\v
VMODULES: OK, value: C:\Users\AndreLuiz\.vmodules
VTMP: OK, value: C:\Users\AndreLuiz\AppData\Local\Temp\v_0

Git version: git version 2.38.1.windows.1
Git vroot status: weekly.2023.13-25-g6ff1c0a0
.git/config present: true

CC version: cc (MinGW-W64 x86_64-msvcrt-posix-seh, built by Brecht Sanders) 12.2.0
thirdparty/tcc status: thirdparty-windows-amd64 1e6e7c6f-dirty

--- a/vlib/v/gen/c/if.v
+++ b/vlib/v/gen/c/if.v
@@ -37,6 +37,16 @@ fn (mut g Gen) need_tmp_var_in_expr(expr ast.Expr) bool {
 		return true
 	}
 	match expr {
+		ast.Ident {
+			return expr.or_expr.kind != .absent
+		}
+		ast.StringInterLiteral {
+			for e in expr.exprs {
+				if g.need_tmp_var_in_expr(e) {
+					return true
+				}
+			}
+		}
 		ast.IfExpr {
 			if g.need_tmp_var_in_if(expr) {
 				return true
@@ -134,7 +144,10 @@ fn (mut g Gen) need_tmp_var_in_expr(expr ast.Expr) bool {
 			}
 		}
 		ast.SelectorExpr {
-			return g.need_tmp_var_in_expr(expr.expr)
+			if g.need_tmp_var_in_expr(expr.expr) {
+				return true
+			}
+			return expr.or_block.kind != .absent
 		}
 		else {}
 	}

--- a/vlib/v/gen/c/match.v
+++ b/vlib/v/gen/c/match.v
@@ -81,7 +81,7 @@ fn (mut g Gen) match_expr(node ast.MatchExpr) {
 			''
 		}
 		cond_var = g.new_tmp_var()
-		g.write('${g.typ(node.cond_type)} ${cond_var} = ')
+		g.write('${g.typ(node.cond_type)} /*A*/ ${cond_var} = ')
 		g.expr(node.cond)
 		g.writeln(';')
 		g.set_current_pos_as_last_stmt_pos()

--- a/vlib/v/tests/option_match_expr_test.v
+++ b/vlib/v/tests/option_match_expr_test.v
@@ -1,0 +1,30 @@
+pub struct ParamPrecos {
+pub:
+	codigo            string [required]
+	tabela_referencia ?i64
+}
+
+fn test_none() {
+	pp := ParamPrecos{
+		codigo: 'opa'
+	}
+
+	tx_ref := match pp.tabela_referencia {
+		none { 'num: none' }
+		else { 'num: ${pp.tabela_referencia?}' }
+	}
+	assert tx_ref == 'num: none'
+}
+
+fn test_not_none() {
+	pp := ParamPrecos{
+		codigo: 'opa'
+		tabela_referencia: 123
+	}
+
+	tx_ref := match pp.tabela_referencia {
+		none { 'num: none' }
+		else { 'num: ${pp.tabela_referencia?}' }
+	}
+	assert tx_ref == 'num: 123'
+}

--- a/vlib/v/tests/option_match_expr_test.v
+++ b/vlib/v/tests/option_match_expr_test.v
@@ -1,30 +1,30 @@
 pub struct ParamPrecos {
 pub:
-	codigo            string [required]
-	tabela_referencia ?i64
+	code      string [required]
+	table_ref ?i64
 }
 
 fn test_none() {
 	pp := ParamPrecos{
-		codigo: 'opa'
+		code: 'V'
 	}
 
-	tx_ref := match pp.tabela_referencia {
+	tx_ref := match pp.table_ref {
 		none { 'num: none' }
-		else { 'num: ${pp.tabela_referencia?}' }
+		else { 'num: ${pp.table_ref?}' }
 	}
 	assert tx_ref == 'num: none'
 }
 
 fn test_not_none() {
 	pp := ParamPrecos{
-		codigo: 'opa'
-		tabela_referencia: 123
+		code: 'V'
+		table_ref: 123
 	}
 
-	tx_ref := match pp.tabela_referencia {
+	tx_ref := match pp.table_ref {
 		none { 'num: none' }
-		else { 'num: ${pp.tabela_referencia?}' }
+		else { 'num: ${pp.table_ref?}' }
 	}
 	assert tx_ref == 'num: 123'
 }

